### PR TITLE
10.0 mig sale order type

### PR DESCRIPTION
**Pull request by PierreFaniel**

### CODE REVIEW

**Comment by PierreFaniel** [comment]
Hello community,

In a previous [pull request](https://github.com/OCA/sale-workflow/pull/358) I received a comment saying that there were changes between the base I used and the current 9.0 version and that I had to follow the [migration to v10 guide](https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-10.0#technical-method-to-migrate-a-module-from-90-to-100-branch)
Which is what I did with this new pull request. 

But now the diff is too big between the 10.0 branch (which I believe is based on the v8.0) and my pull request to pass the code coverage checks.

Does someone have a solution ?

**Comment by pedrobaeza** [comment]
Well, the coverage difference is not so big (only 7%). You can solve this adding a few tests or it can be considered if the drop is acceptable.

**Comment by PierreFaniel** [comment]
Thanks for the fast answer !
I'll check what I can do with the testing.

**Comment by Giuliano69** [comment]
Hi Pierre,
I would like to test the module, BUT I cannot find any Niboo:10.0-mig-sale_order_type under 
https://github.com/Niboo/niboo-community
nor under your repo
https://github.com/PierreFaniel

WHERE is located your PR ?

**Comment by PierreFaniel** [comment]
Hello @Giuliano69 

The PR is on https://github.com/Niboo/sale-workflow/tree/10.0-mig-sale_order_type

Cheers

**Comment by pedrobaeza** [comment]
You should try on runbot. See https://odoo-community.org/page/review

**Review by eLBati — APPROVED** [review]
Thanks!
Could you just squash the last 2 commits in 1 ?

**Comment by PierreFaniel** [comment]
@eLBati sure

**Comment by Giuliano69** [comment]
Hi Pedro,
The runbot running seems not to have the module installed.

May be the runbot
http://3277920-10-0-1a44b0.runbot1.odoo-community.org/web?debug=1#id=411&view_type=form&model=ir.module.module&menu_id=52&action=37
has NOT  sale_order_type module installed nor installable,
due to lack of module stock and sale_stock ?

The module  lackness  can be fixed as an odoo user or as admin ?

**Comment by pedrobaeza** [comment]
You don't have to try in the main runbot, but the runbot link inside this PR. Click on "Show all checks" and on "Details" in the ci/runbot line.

**Comment by eLBati** [comment]
@PierreFaniel please consider making 'sale order type' field (on partner) company dependent. See https://github.com/OCA/sale-workflow/pull/488

File: sale_order_type/models/sale_order_type.py [hunk 1]
```diff
@@ -25,43 +19,21 @@ def default_picking_policy(self):
...
-
-    def default_invoice_state(self):
-        default_dict = self.env['stock.picking'].default_get(['invoice_state'])
-        return default_dict.get('invoice_state')
-
     name = fields.Char(string='Name', required=True, translate=True)
     description = fields.Text(string='Description', translate=True)
     sequence_id = fields.Many2one(
-        comodel_name='ir.sequence', string='Entry Sequence', copy=False,
-        domain=_get_domain_sequence_id)
```
Comment (rgarnau) [inline sale_order_type/models/sale_order_type.py]: Hi @PierreFaniel - how did you migrate this domain?

@aheficent @mreficent

**Comment by pedrobaeza** [comment]
Please include https://github.com/OCA/sale-workflow/pull/488 in this migration (without the migration script).

**Review by rgarnau — CHANGES_REQUESTED** [review]

File: sale_order_type/README.rst [hunk 2]
```diff
@@ -16,7 +20,7 @@ Usage
```
Comment (cubells) [inline sale_order_type/README.rst:27]: Can you add some information for normal users? Something like:

Configuration
=============

To configure this module, you need to:

#. Go to *Sales > Configuration > Sales Orders Types*.
#. Define a new sale order type with all settings you want.

File: sale_order_type/README.rst [hunk 3]
```diff
@@ -16,7 +20,7 @@ Usage
 
 .. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
```
Comment (cubells) [inline sale_order_type/README.rst]: And in Usage section you can add something like:

#. Go to *Sales > Sales Orders* and create a new one.
#. Select the new type you are created before and all settings will be 
   propagated.
#. Or you can define a type for a particular partner if you go to *Sales &
   Purchases* and set a sale order type.

File: sale_order_type/security/security.xml [hunk 4]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0"?>
+<openerp>
```
Comment (cubells) [inline sale_order_type/security/security.xml]: change to odoo

File: sale_order_type/security/security.xml [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    <data noupdate="0">
```
Comment (cubells) [inline sale_order_type/security/security.xml]: remove data

File: sale_order_type/views/account_invoice_view.xml [hunk 6]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
```
Comment (cubells) [inline sale_order_type/views/account_invoice_view.xml]: change to odoo

File: sale_order_type/views/account_invoice_view.xml [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+    <data>
```
Comment (cubells) [inline sale_order_type/views/account_invoice_view.xml]: remove data tag as is is indicated in this page: https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-10.0

**Review by cubells — CHANGES_REQUESTED** [review]
Some minor changes.

File: sale_order_type/data/default_type.xml [hunk 8]
```diff
@@ -1,10 +1,10 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<openerp>
+<odoo>
```
Comment (tafaRU) [inline sale_order_type/data/default_type.xml]: you can move `noupdate="1"` here, I mean at `<odoo>` tag level.
In addition, please do the same for the file `demo/sale_order_demo.xml` after to have replaced tag `openerp` by `odoo`.

File: sale_order_type/models/__init__.py [hunk 9]
```diff
@@ -1,11 +1,9 @@
 # -*- encoding: utf-8 -*-
 ##############################################################################
```
Comment (tafaRU) [inline sale_order_type/models/__init__.py]: It should be better to use https://github.com/OCA/maintainer-tools/blob/master/template/module/models/__init__.py#L1-L2

File: sale_order_type/models/account_invoice.py [hunk 10]
```diff
@@ -1,30 +1,31 @@
 # -*- encoding: utf-8 -*-
 ##############################################################################
```
Comment (tafaRU) [inline sale_order_type/models/account_invoice.py]: Please use https://github.com/OCA/maintainer-tools/blob/master/template/module/models/model_name.py#L1-L3

File: sale_order_type/models/res_partner.py [hunk 11]
```diff
@@ -1,13 +1,14 @@
 # -*- encoding: utf-8 -*-
```
Comment (tafaRU) [inline sale_order_type/models/res_partner.py]: Same as above

File: sale_order_type/models/sale_order.py [hunk 12]
```diff
@@ -1,56 +1,55 @@
 # -*- encoding: utf-8 -*-
```
Comment (tafaRU) [inline sale_order_type/models/sale_order.py]: Same as above.

File: sale_order_type/models/sale_order.py [hunk 13]
```diff
@@ -1,56 +1,55 @@
...
+                order.payment_term_id = order.type_id.payment_term_id.id
+            if order.type_id.pricelist_id:
+                order.pricelist_id = order.type_id.pricelist_id.id
+            if order.type_id.incoterm_id:
+                order.incoterm = order.type_id.incoterm_id.id
 
     @api.model
     def create(self, vals):
         if vals.get('name', '/') == '/'and vals.get('type_id'):
             type = self.env['sale.order.type'].browse(vals['type_id'])
```
Comment (tafaRU) [inline sale_order_type/models/sale_order.py]: It should be better don't use built-in function name for your variable. Please rename it to `sale_type`.

File: sale_order_type/models/sale_order_type.py [hunk 14]
```diff
@@ -1,9 +1,9 @@
 # -*- encoding: utf-8 -*-
```
Comment (tafaRU) [inline sale_order_type/models/sale_order_type.py]: Same as above.

**Review by tafaRU — CHANGES_REQUESTED** [review]

**Review by cubells — APPROVED** [review]
:+1: 

LGTM

**Comment by tafaRU** [comment]
@PierreFaniel, thanks a lot!
Last remark: please squash a bit the commit history as mentioned at https://github.com/OCA/maintainer-tools/wiki/Merge-commits-in-pull-requests expecially at https://github.com/OCA/maintainer-tools/wiki/Merge-commits-in-pull-requests#mergesquash-the-commits-generated-by-transifex

**Comment by PierreFaniel** [comment]
Should I squash them all into one ? Wouldn't that be bad to remove all the history ?
Or do you mean squash my commits into one ?

**Comment by MiquelRForgeFlow** [comment]
He refers to squash adjacent commits whenever they share same author or aren't important additions.

**Comment by tafaRU** [comment]
@PierreFaniel, at least, I'd squash together Transifex commits as mentioned at https://github.com/OCA/maintainer-tools/wiki/Merge-commits-in-pull-requests#mergesquash-the-commits-generated-by-transifex

Thanks!

**Comment by cubells** [comment]
@PierreFaniel :
your module tests are failing. You must review...

**Review by tafaRU — APPROVED** [review]

**Review by MiquelRForgeFlow — APPROVED** [review]
code review :+1:

File: sale_order_type/views/sale_order_type_view.xml [hunk 15]
```diff
@@ -1,119 +1,114 @@
...
-                            </group>
-                            <group>
-                                <field name="journal_id" />
-                                <field name="refund_journal_id" />
-                                <field name="picking_policy" />
-                                <field name="order_policy" />
-                                <field name="invoice_state" />
-                            </group>
+                            <field name="warehouse_id" />
+                            <field name="company_id" />
```
Comment (MiquelRForgeFlow) [inline sale_order_type/views/sale_order_type_view.xml]: I think it should be nicer if `company_id` has the attribute `invisible=true` whenever `warehouse_is` is empty.

File: sale_order_type/__manifest__.py [hunk 16]
```diff
@@ -1,51 +1,38 @@
...
+# -*- coding: utf-8 -*-
+# Copyright 2015 Carlos Sánchez Cifuentes <csanchez@grupovermon.com>
+# Copyright 2015-2016 Oihane Crucelaegui <oihane@avanzosc.com>
+# Copyright 2015-2016 Pedro M. Baeza <pedro.baeza@tecnativa.com>
+# Copyright 2016 Lorenzo Battistini
+# Copyright 2016 Carlos Dauden <carlos.dauden@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
 
 {
     "name": "Sale Order Types",
```
Comment (MiquelRForgeFlow) [inline sale_order_type/__manifest__.py]: the real name of the module is `Sale Order Type`, and this is what it should be shown

**Review by MiquelRForgeFlow — CHANGES_REQUESTED** [review]
Tested on runbot :+1:

**Review by MiquelRForgeFlow — APPROVED** [review]

File: sale_order_type/models/sale_order.py [hunk 17]
```diff
@@ -1,56 +1,53 @@
...
-            partner = self.env['res.partner'].browse(part)
-            res['value'].update({
-                'type_id': partner.sale_type.id or self._get_order_type().id,
-            })
-        return res
+    @api.onchange('partner_id')
+    def onchange_partner_id(self):
+        super(SaleOrder, self).onchange_partner_id()
+        if self.partner_id:
+            self.type_id = self.partner_id.sale_type or self._get_order_type()
```
Comment (carlosdauden) [inline sale_order_type/models/sale_order.py]: Change this to don't overwrite user default value:

`        if self.partner_id.sale_type:
            self.type_id = self.partner_id.sale_type
`

**Review by carlosdauden — CHANGES_REQUESTED** [review]

File: sale_order_type/models/sale_order.py (hunk 17 above)
Comment (carlosdauden) [reply sale_order_type/models/sale_order.py]: self._get_order_type() is called by default. It shouldn't be recalled at the onchange_partner

File: sale_order_type/models/sale_order.py [hunk 18]
```diff
@@ -1,56 +1,53 @@
...
-        if part:
-            partner = self.env['res.partner'].browse(part)
-            res['value'].update({
-                'type_id': partner.sale_type.id or self._get_order_type().id,
-            })
-        return res
+    @api.onchange('partner_id')
+    def onchange_partner_id(self):
+        super(SaleOrder, self).onchange_partner_id()
+        if self.partner_id and self.partner_id.sale_type:
```
Comment (carlosdauden) [inline sale_order_type/models/sale_order.py]: In API 8.0 it is not necessary to check if partner_id is set

**Review by carlosdauden — APPROVED** [review]
:+1: 
Minnor comments

**Review by rgarnau — APPROVED** [review]

File: sale_order_type/__manifest__.py [hunk 19]
```diff
@@ -1,51 +1,38 @@
...
-              "Grupo Vermon, "
-              "AvanzOSC, "
-              "Serv. Tecnol. Avanzados - Pedro M. Baeza, "
+    "author": "Grupo Vermon,"
+              "AvanzOSC,"
+              "Tecnativa,"
+              "Agile Business Group,"
+              "Niboo,"
               "Odoo Community Association (OCA)",
     "website": "http://www.odoomrp.com",
```
Comment (AaronHForgeFlow) [inline sale_order_type/__manifest__.py]: I'd use https://github.com/OCA/sale-workflow and also can you add a short summary?

File: sale_order_type/models/account_invoice.py [hunk 20]
```diff
@@ -1,30 +1,29 @@
-# -*- encoding: utf-8 -*-
-##############################################################################
-# For copyright and license notices, see __openerp__.py file in root directory
-##############################################################################
+# -*- coding: utf-8 -*-
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
 
-from openerp import api, models
+from odoo import api, models, fields
```
Comment (AaronHForgeFlow) [inline sale_order_type/models/account_invoice.py]: from odoo import api, fields, models

**Review by AaronHForgeFlow — CHANGES_REQUESTED** [review]
Hi! @PierreFaniel Thank you for this migration. It works like a charm :fire:  Just some minor comments.

**Review by AaronHForgeFlow — APPROVED** [review]
:+1: Thank you for attending all the comments

**Comment by PierreFaniel** [comment]
NP, thank you all for your thorough reviews !

**Comment by ianathompson** [comment]
Hello @PierreFaniel The module for sale_order_type is working well for us, but I did find one issue.
When parsing or Creating a new Sales Order Sequence, all Sequence Codes **must** be named `sale.order` if you want to create a New Sales Order Sequence, and give it a difference sequence code maybe `sale.order.region1` it will not show up on the drop down list in the Create New Sales Order Type for Sequence. Additionally, if you leave the `Sequence Code` blank, it will also not display.
If you create a new one without a sequence code of sale.order it will still list on the Sales Order Type, but it will not be queried in the drop-down list.
Not sure how to fix this, or if this is by design. If the sequence code must always be sale.order for every sales_order_type that is setup, then something should be mentioned in the readme regarding this.

**Pull request merged**